src/transmuxer_worker.js

```diff
@@ -17,39 +17,100 @@ importScripts('../node_modules/mux.js/lib/metadata-stream.js');
 importScripts('../node_modules/mux.js/lib/transmuxer.js');
 importScripts('../node_modules/mux.js/lib/caption-stream.js');
 
-var transmuxer = new muxjs.mp2t.Transmuxer();
+var transmuxer;
 
-onmessage = function(event) {
-  if (event.data.action === 'push') {
-    // Cast to type
-    var segment = new Uint8Array(event.data.data);
+/**
+ * wireTransmuxerEvents
+ * Re-emits tranmsuxer events by converting them into messages to the
+ * world outside the worker
+ */
+var wireTransmuxerEvents = function (transmuxer) {
+  transmuxer.on('data', function (segment) {
+    // transfer ownership of the underlying ArrayBuffer instead of doing a copy to save memory
+    // ArrayBuffers are transferable but generic TypedArrays are not
+    // see https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers#Passing_data_by_transferring_ownership_(transferable_objects)
+    segment.data = segment.data.buffer;
+    postMessage({
+      action: 'data',
+      segment: segment
+    }, [segment.data]);
+  });
+
+  if (transmuxer.captionStream) {
+    transmuxer.captionStream.on('data', function(caption) {
+      postMessage({
+        action: 'caption',
+        data: caption
+      });
+    });
+  }
+
+  transmuxer.on('done', function (data) {
+    postMessage({ action: 'done' });
+  });
+};
 
+/**
+ * All incoming messages route through this hash. If not function exists
+ * to handle an incoming message, then we ignore the message.
+ */
+var messageHandlers = {
+  /**
+   * init
+   * Allows you to initialize the transmuxer and pass along options from
+   * outside the worker
+   */
+  init: function (data) {
+    transmuxer = new muxjs.mp2t.Transmuxer(data && data.options);
+    wireTransmuxerEvents(transmuxer);
+  },
+  /**
+   * defaultInit
+   * Is called before every function and initializes the transmuxer with
+   * default options if `init` was never explicitly called
+   */
+  defaultInit: function () {
+    if (!transmuxer) {
```

Comment:
What if this was just moved into `onmessage` below? Then we wouldn't have to remember to call it in every future message handler we create.